### ISSUE TITLE
Parallel TensorFlow Model Training Ensemble from Streaming Data

### ISSUE THREAD

**Issue by samjabrahams** [fluxcapacitor/pipeline#114]
Basic specs:
- Single point of entry for client
- Manager keeps track of key/model-family pairings
- "Family" of models represent models with the same graph but different parameters/hyperparameters
- Each family has "Trainers" responsible for the lifecycle of individual models
- For now, write out TensorBoard/Model checkpoints directly to disk
- For now, models run on same machine

**Comment by cfregly** [comment]
More notes:

Items to be versioned together:
1) TensorFlow Code
2) Model/graph definition (shape of model)
3) State of hyper-parameters for each training batch (potentially not store at every training batch)
4) TensorBoard summary writer at each training batch:  loss graph, histrogram of variables, etc

in a real-time streaming context, kafka will buffer any spikes in traffic throughout the day.  batch sizes can stay the same for each experiment - versus subject to trends.

**Comment by cfregly** [comment]
Possible solutions:
* Mutably
* Immutable branching

Reminder to initialize all parallel models with the same initialization weights otherwise parallel experiments are invalid. 

Tie back to kafka statistics like input rate, etc.

**TIME**
Need to think about kafka's concept of time versus tensorflow's concept of time.  Because we're capping TensorFlow's batch size (ie 64), TensorFlow may not see all of the data that has come into the Kafka topic.  We might have to drain the Kafka topic before we can write the TensorFlow version/snapshot/SummaryWriter.

**Comment by cfregly** [comment]
Related to https://github.com/fluxcapacitor/pipeline/issues/50

**Comment by cfregly** [comment]
examples of image serialization in and out of kafka:  https://github.com/deeplearning4j/dl4j-streaming/tree/master/src/main/java/org/deeplearning4j/streaming/serde

**Comment by cfregly** [comment]
learning rate should scale along with batch size (smaller batch -> smaller learning rate).

for more context:  ideally, if we had all data (batch gradient descent), we would always find global minimum by traversing all data in a single epoch.

**Comment by cfregly** [comment]
tf.nn.batch_normalize is key!  continuously learns and updates std_dev() and mean() for input data as new batches are seen.  these learned values (stddev and mean) are saved with the model.

**Comment by cfregly** [comment]
Setup Distributed Tensorboard using HDFS

**Comment by cfregly** [comment]
Part of Larger Effort.  Moving to [PipelineAI Advanced/Enterprise Edition](http://pipeline.ai/products).
